services/frontend/server.go

```diff
@@ -149,3 +158,75 @@ func (s *server) handleReload(logger zerolog.Logger) http.HandlerFunc {
 		w.WriteHeader(http.StatusOK)
 	}
 }
+
+func (s *server) watchConfigs(logger zerolog.Logger) {
+	s.group.Add(1)
+	defer s.group.Done()
+
+	ticker := time.NewTicker(40 * time.Second)
+
+	// keep track of which configs are done so we don't endlessly print the status.
+	inProgress := make(map[string]*pb.Config)
```

Comment:
We should rename this map since it is holding configs which are done, not in progress.

EDIT: After reading the code more, I think you meant to have `!=` with `&&` when you are assigning entries to map, like
```go
for _, cfg := range resp.GetConfigs() {
		for cluster, wf := range cfg.State {
			if wf.Status != pb.Workflow_ERROR  && wf.Status != pb.Workflow_DONE {
				inProgress[cluster] = cfg
			}
		}
	}
```

Reply:
The map is actually holding configs that are in progress and once they error or are done only a single message is printed and then deleted from the map


```go
for _, cfg := range resp.GetConfigs() {
		for cluster, wf := range cfg.State {
			if wf.Status == pb.Workflow_ERROR || wf.Status == pb.Workflow_DONE {
				inProgress[cluster] = cfg
			}
		}
	}
```

So this is actually just to save configs that errored or finished so we can print the messages, `workflow for cluster ... finished/failed ...`

I added this code so that when you start the frontend and there are some configs that finished or error so atleast one message will get logged